core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import zio.Synchronizer._
+
+final class CountDownLatch(count: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(count, maxWaiters)
+
+  // TODO is it possible to evaluate once and reuse the Promise?
+  private[this] val awaitZero = sync.acquire(count)
+
+  def await: UIO[Unit] = awaitZero.asInstanceOf[UIO[Unit]]
+
+  // TODO stop at zero
+  def countDown(): UIO[Unit] = sync.release(1)
+}
+
+final class AsyncSemaphore(initialPermits: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(initialPermits, maxWaiters)
+
+  def apply[R, E, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    sync
+      .acquire(1)
+      .mapError(Right(_))
+      .flatMap { _ =>
+        zio
+          .mapError(Left(_))
+          .ensuring(sync.release(1))
+      }
+}
+
+final class ReentrantLock(maxWaiters: Option[Int]) {
+  @volatile private[this] var owner: Fiber.Id = null
+
+  private[this] val sem = new AsyncSemaphore(1, maxWaiters)
+
+  def apply[R, E <: Throwable, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    ZIO.effectSuspendTotalWith { (_, fiberId) =>
+      if (owner == fiberId) {
+        zio.mapError(Left(_))
+      } else {
+        sem(ZIO.effectSuspendTotal {
+          owner = fiberId
+          zio.ensuring(ZIO.effectTotal(owner = null))
+        })
+      }
+    }
+}
+
+final class Synchronizer(initialPermits: Int, maxWaiters: Int) {
```

Comment:
`acquire` and `release` are missing validations in case the caller misuses the API. I think this class should be `private[zio]`, though